activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/service/ModelService.java

```diff
@@ -63,277 +68,391 @@
-        return modelRepository.updateModel(modelToBeUpdated,
-                                           newModel);
+    Model fullModel = findModelById(model.getId()).orElse(model);
+    byte[] modelContent = modelRepository.getModelContent(model);
+    final String bpmnModelId = modelContentService
+      .findModelContentConverter(model.getType())
+      .flatMap(converter -> converter.convertToModelContent(modelContent))
+      .map(ModelContent::getId)
+      .orElseGet(() -> modelContentService.getModelContentId(model));
+
+    Model modelToFile = buildModel(fullModel.getType(), fullModel.getName());
+    modelToFile.setId(bpmnModelId);
+    modelToFile.setExtensions(fullModel.getExtensions());
+
+    FileContent metadataFileContent = new FileContent(getMetadataFilename(model),
+      CONTENT_TYPE_JSON,
+      jsonConverter.convertToJsonBytes(modelToFile));
+    return Optional.of(metadataFileContent);
+  }
+
+  public void cleanModelIdList(){
+    this.modelIdentifiers.clear();
+  }
+
+  public Optional<FileContent> getModelDiagramFile(String modelId) {
+    //TODO: to implement
+    return Optional.empty();
+  }
+
+  public String getMetadataFilename(Model model) {
+    return toJsonFilename(model.getName() + findModelType(model).getMetadataFileSuffix());
+  }
+
+  public FileContent getModelContentFile(Model model) {
+    return getModelFileContent(model,
+      modelRepository.getModelContent(model));
+  }
+
+  public FileContent exportModel(Model model) {
+    return getModelFileContent(model,
+      modelRepository.getModelExport(model));
+  }
+
+  private FileContent getModelFileContent(Model model,
+                                          byte[] modelBytes) {
+    return new FileContent(setExtension(model.getName(),
+      findModelType(model).getContentFileExtension()),
+      model.getContentType(),
+      modelBytes);
+  }
+
+  public Model updateModelContent(Model modelToBeUpdate,
+                                  FileContent fileContent) {
+    FileContent fixedFileContent = this.modelIdentifiers.isEmpty()?
+      fileContent:
+      overrideModelContentId(modelToBeUpdate, fileContent);
+
+    modelToBeUpdate.setContentType(fixedFileContent.getContentType());
+    modelToBeUpdate.setContent(fixedFileContent.toString());
+
+    Optional.ofNullable(modelToBeUpdate.getType())
+      .flatMap(modelContentService::findModelContentConverter)
+      .flatMap(validator -> validator.convertToModelContent(fixedFileContent.getFileContent()))
+      .ifPresent(modelContent -> modelToBeUpdate.setTemplate(modelContent.getTemplate()));
+
+    return modelRepository.updateModelContent(modelToBeUpdate, fixedFileContent);
+  }
+
+
+  public FileContent overrideModelContentId(Model model, FileContent fileContent) {
+    FileContent fixedFileContent = null;
+    switch (model.getType()){
+      case "PROCESS": {
+        fixedFileContent = this.overridingProcessContentId(model, fileContent);
+        break;
+      }
+      case "FORM":{
```

Comment:
This is enterprise code, it should go to alfresco-process-modeling. There is a dedicated FormModelContentConverter which is responsible for this kind of operations. This way you can keep the generics clean in ModelContentConverter and strongly control the model types you are working on. I suggest to keep layers separation, otherwise this ModelService will become very hard to maintain and understand. We can discuss about...